File: esports-backend/src/main/java/com/esports/service/UserService.java
package com.esports.service;

import com.esports.dto.*;
import com.esports.entity.User;
import com.esports.entity.UserProfile;
import com.esports.repository.UserRepository;
import com.esports.repository.UserProfileRepository;
import com.esports.security.JwtUtil;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken;
import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.gson.GsonFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.Collections;
import java.util.NoSuchElementException;
import java.util.Optional;

@Service
public class UserService implements UserDetailsService {

    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private UserProfileRepository userProfileRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private JwtUtil jwtUtil;

    @Value("${spring.security.oauth2.client.registration.google.client-id}")
    private String googleClientId;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        System.out.println("Register request: " + request); // Debug log
        
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("Email already exists");
        }

        User user = new User();
        user.setName(request.getName());
        user.setEmail(request.getEmail());
        user.setPassword(passwordEncoder.encode(request.getPassword()));
        
        try {
            if (request.getEmail().contains("@admin")) {
                user.setRole(User.Role.ADMIN);
            } else if (request.getRole() != null && !request.getRole().trim().isEmpty()) {
                user.setRole(User.Role.valueOf(request.getRole().trim()));
            } else {
                throw new IllegalArgumentException("Role is required");
            }
        } catch (IllegalArgumentException e) {
            if (e.getMessage().contains("Role is required")) {
                throw e;
            }
            throw new IllegalArgumentException("Invalid role: " + request.getRole());
        }

        user = userRepository.save(user);
        user.setHasSelectedRole(true);
        user = userRepository.save(user);
        String token = jwtUtil.generateToken(user.getEmail());

        return new AuthResponse(user.getId(), user.getName(), user.getEmail(), user.getRole(), token, user.getCreatedAt());
    }

    public AuthResponse login(LoginRequest request) {
        Optional<User> userOpt = userRepository.findByEmail(request.getEmail());
        
        if (userOpt.isEmpty()) {
            throw new IllegalArgumentException("No account found with this email address");
        }
        
        User user = userOpt.get();
        
        // (Google OAuth user)
        if (user.getGoogleId() != null) {
            throw new IllegalArgumentException("Please use Google Sign-In for this account");
        }
        
        // Check password for regular users
        if (user.getPassword() == null || !passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new IllegalArgumentException("Incorrect password");
        }
        
        if (!user.isActive()) {
            throw new IllegalArgumentException("Account has been deactivated");
        }
        
        String token = jwtUtil.generateToken(user.getEmail());

        return new AuthResponse(user.getId(), user.getName(), user.getEmail(), user.getRole(), token, user.getCreatedAt());
    }

    public AuthResponse googleAuth(GoogleAuthRequest request) {
        try {
            GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(
                new NetHttpTransport(), GsonFactory.getDefaultInstance())
                .setAudience(Collections.singletonList(googleClientId))
                .build();

            GoogleIdToken idToken;
            try {
                idToken = verifier.verify(request.getToken());
            } catch (Exception verificationError) {
                throw new RuntimeException("Google authentication failed");
            }
            
            if (idToken == null) {
                throw new RuntimeException("Invalid Google token");
            }

            GoogleIdToken.Payload payload = idToken.getPayload();
            String googleId = payload.getSubject();
            String email = payload.getEmail();
            String name = (String) payload.get("name");

            Optional<User> existingUser = userRepository.findByGoogleId(googleId);
            User user;

            if (existingUser.isPresent()) {
                user = existingUser.get();
                if (!user.isActive()) {
                    throw new RuntimeException("Account has been deactivated");
                }
            } else {
                Optional<User> emailUser = userRepository.findByEmail(email);
                if (emailUser.isPresent()) {
                    user = emailUser.get();
                    if (!user.isActive()) {
                        throw new RuntimeException("Account has been deactivated");
                    }
                    user.setGoogleId(googleId);
                    user = userRepository.save(user);
                } else {
                    user = new User();
                    user.setName(name);
                    user.setEmail(email);
                    user.setGoogleId(googleId);
                    user.setHasSelectedRole(false);
                    
                    if (email.contains("@admin")) {
                        user.setRole(User.Role.ADMIN);
                        user.setHasSelectedRole(true);
                    } else {
                        user.setRole(User.Role.CREATOR);
                    }
                    user = userRepository.save(user);
                }
            }

            String token = jwtUtil.generateToken(user.getEmail());
            return new AuthResponse(user.getId(), user.getName(), user.getEmail(), user.getRole(), token, user.getCreatedAt(), user.isHasSelectedRole());

        } catch (Exception e) {
            throw new RuntimeException(e.getMessage());
        }
    }

    @Override
    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new UsernameNotFoundException("User not found: " + email));

        return new org.springframework.security.core.userdetails.User(
            user.getEmail(),
            user.getPassword() != null ? user.getPassword() : "",
            Collections.singletonList(new SimpleGrantedAuthority("ROLE_" + user.getRole().name()))
        );
    }

    public UserResponseDTO getCurrentUser(String email) {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        UserProfile profile = userProfileRepository.findByUser(user).orElse(null);
        
        UserResponseDTO response = new UserResponseDTO();
        response.setId(user.getId());
        response.setName(user.getName());
        response.setEmail(user.getEmail());
        response.setRole(user.getRole().name());
        response.setCreatedAt(user.getCreatedAt());
        
        if (profile != null) {
            response.setBio(profile.getBio());
            response.setLocation(profile.getLocation());
            response.setWebsite(profile.getWebsite());
            response.setSkills(profile.getSkills());
            response.setProfessionalName(profile.getProfessionalName());
            response.setChannelName(profile.getChannelName());
        }
        
        return response;
    }

    @Transactional
    public UserResponseDTO updateUserProfile(String email, UserProfileUpdateDTO updateData) {
        User user = userRepository.findByEmail(email)
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        if (updateData.getName() != null) {
            user.setName(updateData.getName());
        }
        
        // email updates for admin users
        if (updateData.getEmail() != null && !updateData.getEmail().equals(user.getEmail())) {
            // Check if user is admin 
            if (user.getRole() == User.Role.ADMIN) {
                // Check if email already exists
                if (userRepository.existsByEmail(updateData.getEmail())) {
                    throw new IllegalArgumentException("Email already exists");
                }
                user.setEmail(updateData.getEmail());
            }
        }
        
        User updatedUser = userRepository.save(user);
        
        UserProfile profile = userProfileRepository.findByUser(user)
            .orElse(new UserProfile());
        
        if (profile.getUser() == null) {
            profile.setUser(user);
        }
        
        if (updateData.getBio() != null) {
            profile.setBio(updateData.getBio());
        }
        if (updateData.getLocation() != null) {
            profile.setLocation(updateData.getLocation());
        }
        if (updateData.getWebsite() != null) {
            profile.setWebsite(updateData.getWebsite());
        }
        if (updateData.getSkills() != null) {
            profile.setSkills(updateData.getSkills());
        }
        if (updateData.getProfessionalName() != null) {
            profile.setProfessionalName(updateData.getProfessionalName());
        }
        if (updateData.getChannelName() != null) {
            profile.setChannelName(updateData.getChannelName());
        }
        
        UserProfile updatedProfile = userProfileRepository.save(profile);
        
        UserResponseDTO response = new UserResponseDTO();
        response.setId(updatedUser.getId());
        response.setName(updatedUser.getName());
        response.setEmail(updatedUser.getEmail());
        response.setRole(updatedUser.getRole().name());
        response.setCreatedAt(updatedUser.getCreatedAt());
        response.setBio(updatedProfile.getBio());
        response.setLocation(updatedProfile.getLocation());
        response.setWebsite(updatedProfile.getWebsite());
        response.setSkills(updatedProfile.getSkills());
        response.setProfessionalName(updatedProfile.getProfessionalName());
        response.setChannelName(updatedProfile.getChannelName());
        
        return response;
    }

    @Transactional
    public AuthResponse updateUserRole(UpdateRoleRequest request) {
        User user = userRepository.findById(request.getUserId())
            .orElseThrow(() -> new NoSuchElementException("User not found"));
        
        if (user.getGoogleId() == null) {
            throw new IllegalArgumentException("Role update only allowed for Google users");
        }
        
        try {
            user.setRole(User.Role.valueOf(request.getRole()));
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid role: " + request.getRole());
        }
        
        user = userRepository.save(user);
        user.setHasSelectedRole(true);
        user = userRepository.save(user);
        String token = jwtUtil.generateToken(user.getEmail());
        
        return new AuthResponse(user.getId(), user.getName(), user.getEmail(), user.getRole(), token, user.getCreatedAt());
    }
}

File: Upkart1/Services/UserService.cs
using Microsoft.AspNetCore.Identity;
using Microsoft.EntityFrameworkCore;
using Google.Apis.Auth;
using Upkart1.Data;
using Upkart1.DTO;
using Upkart1.Entities;
using Upkart1.Entities.Enums;
using Upkart1.Security;

namespace Upkart1.Services
{
    public class UserService : IUserService
    {
        private readonly ApplicationDbContext _context;
        private readonly IPasswordHasher<User> _passwordEncoder;
        private readonly JwtUtil _jwtUtil;
        private readonly string _googleClientId;

        public UserService(
            ApplicationDbContext context,
            IPasswordHasher<User> passwordEncoder,
            JwtUtil jwtUtil,
            IConfiguration configuration)
        {
            _context = context;
            _passwordEncoder = passwordEncoder;
            _jwtUtil = jwtUtil;
            _googleClientId = configuration["Authentication:Google:ClientId"];
        }

        public AuthResponse Register(RegisterRequest request)
        {
            if (_context.Users.Any(u => u.Email == request.Email))
                throw new ArgumentException("Email already exists");

            var user = new User
            {
                Name = request.Name,
                Email = request.Email
            };

            user.Password = _passwordEncoder.HashPassword(user, request.Password);

            try
            {
                if (request.Email.Contains("@admin"))
                {
                    user.Role = Role.ADMIN;
                }
                else if (!string.IsNullOrWhiteSpace(request.Role))
                {
                    user.Role = Enum.Parse<Role>(request.Role.Trim());
                }
                else
                {
                    throw new ArgumentException("Role is required");
                }
            }
            catch (ArgumentException e) when (e.Message.Contains("Role is required"))
            {
                throw;
            }
            catch (ArgumentException)
            {
                throw new ArgumentException($"Invalid role: {request.Role}");
            }

            _context.Users.Add(user);
            _context.SaveChanges();

            user.HasSelectedRole = true;
            _context.SaveChanges();

            var token = _jwtUtil.GenerateToken(user.Email, user.Role.ToString());

            return new AuthResponse(user.Id, user.Name, user.Email, user.Role, token, user.CreatedAt);
        }

        public AuthResponse Login(LoginRequest request)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == request.Email);
            if (user == null)
                throw new ArgumentException("No account found with this email address");

            if (user.GoogleId != null)
                throw new ArgumentException("Please use Google Sign-In for this account");

            if (user.Password == null || _passwordEncoder.VerifyHashedPassword(user, user.Password, request.Password) != PasswordVerificationResult.Success)
                throw new ArgumentException("Incorrect password");

            if (!user.Active)
                throw new ArgumentException("Account has been deactivated");

            var token = _jwtUtil.GenerateToken(user.Email, user.Role.ToString());

            return new AuthResponse(user.Id, user.Name, user.Email, user.Role, token, user.CreatedAt);
        }

        public AuthResponse GoogleAuth(GoogleAuthRequest request)
        {
            try
            {
                if (string.IsNullOrEmpty(_googleClientId))
                    throw new Exception("Google Client ID not configured");

                GoogleJsonWebSignature.Payload payload;
                try
                {
                    payload = GoogleJsonWebSignature.ValidateAsync(request.Token, new GoogleJsonWebSignature.ValidationSettings
                    {
                        Audience = new[] { _googleClientId }
                    }).Result;
                }
                catch (Exception ex)
                {
                    throw new Exception($"Google token validation failed: {ex.Message}");
                }

                if (payload == null)
                    throw new Exception("Invalid Google token - payload is null");

                var googleId = payload.Subject;
                var email = payload.Email;
                var name = payload.Name;

                var existingUser = _context.Users.FirstOrDefault(u => u.GoogleId == googleId);
                User user;

                if (existingUser != null)
                {
                    user = existingUser;
                    if (!user.Active)
                        throw new Exception("Account has been deactivated");
                }
                else
                {
                    var emailUser = _context.Users.FirstOrDefault(u => u.Email == email);
                    if (emailUser != null)
                    {
                        user = emailUser;
                        if (!user.Active)
                            throw new Exception("Account has been deactivated");
                        user.GoogleId = googleId;
                        _context.SaveChanges();
                    }
                    else
                    {
                        user = new User
                        {
                            Name = name,
                            Email = email,
                            GoogleId = googleId,
                            HasSelectedRole = false,
                            Active = true
                        };

                        if (email.Contains("@admin"))
                        {
                            user.Role = Role.ADMIN;
                            user.HasSelectedRole = true;
                        }
                        else
                        {
                            user.Role = Role.CREATOR;
                        }
                        _context.Users.Add(user);
                        _context.SaveChanges();
                    }
                }

                var token = _jwtUtil.GenerateToken(user.Email, user.Role.ToString());
                return new AuthResponse(user.Id, user.Name, user.Email, user.Role, token, user.CreatedAt, user.HasSelectedRole);
            }
            catch (Exception e)
            {
                throw new Exception(e.Message);
            }
        }

        public AuthResponse UpdateUserRole(UpdateRoleRequest request)
        {
            var user = _context.Users.FirstOrDefault(u => u.Id == request.UserId)
                ?? throw new KeyNotFoundException("User not found");

            if (user.GoogleId == null)
                throw new ArgumentException("Role update only allowed for Google users");

            try
            {
                user.Role = Enum.Parse<Role>(request.Role);
            }
            catch
            {
                throw new ArgumentException($"Invalid role: {request.Role}");
            }

            _context.SaveChanges();
            user.HasSelectedRole = true;
            _context.SaveChanges();

            var token = _jwtUtil.GenerateToken(user.Email, user.Role.ToString());

            return new AuthResponse(user.Id, user.Name, user.Email, user.Role, token, user.CreatedAt);
        }

        public UserResponseDTO GetCurrentUser(string email)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            var profile = _context.UserProfiles.FirstOrDefault(p => p.UserId == user.Id);

            return new UserResponseDTO(
                user.Id,
                user.Name,
                user.Email,
                user.Role.ToString(),
                user.CreatedAt,
                profile?.Bio ?? "",
                profile?.Location ?? "",
                profile?.Website ?? "",
                profile?.Skills ?? "",
                profile?.ProfessionalName ?? "",
                profile?.ChannelName ?? ""
            );
        }

        public UserResponseDTO UpdateUserProfile(string email, UserProfileUpdateDTO updateData)
        {
            var user = _context.Users.FirstOrDefault(u => u.Email == email)
                ?? throw new KeyNotFoundException("User not found");

            if (updateData.Name != null)
                user.Name = updateData.Name;

            if (updateData.Email != null && updateData.Email != user.Email)
            {
                if (user.Role == Entities.Enums.Role.ADMIN)
                {
                    if (_context.Users.Any(u => u.Email == updateData.Email))
                        throw new ArgumentException("Email already exists");
                    user.Email = updateData.Email;
                }
            }

            var updatedUser = user;
            _context.SaveChanges();

            var profile = _context.UserProfiles.FirstOrDefault(p => p.UserId == user.Id)
                ?? new UserProfile();

            if (profile.UserId == 0)
            {
                profile.UserId = user.Id;
                _context.UserProfiles.Add(profile);
            }

            if (updateData.Bio != null)
                profile.Bio = updateData.Bio;
            if (updateData.Location != null)
                profile.Location = updateData.Location;
            if (updateData.Website != null)
            {
                if (!string.IsNullOrWhiteSpace(updateData.Website) && !Uri.IsWellFormedUriString(updateData.Website, UriKind.Absolute))
                {
                    var validationException = new Upkart1.Exceptions.ValidationException("Validation failed");
                    validationException.ValidationErrors["website"] = "Website must be a valid URL";
                    throw validationException;
                }
                profile.Website = updateData.Website;
            }
            if (updateData.Skills != null)
                profile.Skills = updateData.Skills;
            if (updateData.ProfessionalName != null)
                profile.ProfessionalName = updateData.ProfessionalName;
            if (updateData.ChannelName != null)
                profile.ChannelName = updateData.ChannelName;

            _context.SaveChanges();

            return new UserResponseDTO(
                updatedUser.Id,
                updatedUser.Name,
                updatedUser.Email,
                updatedUser.Role.ToString(),
                updatedUser.CreatedAt,
                profile.Bio ?? "",
                profile.Location ?? "",
                profile.Website ?? "",
                profile.Skills ?? "",
                profile.ProfessionalName ?? "",
                profile.ChannelName ?? ""
            );
        }

        public User? LoadUserByUsernameAsync(string email)
        {
            return _context.Users.FirstOrDefault(u => u.Email == email);
        }
    }
}
